### DISCUSSION TITLE
Installing in multi-user environment on HPC

### DISCUSSION THREAD

**Discussion by lloydgoodman** [idaholab/moose#18018] [Q&A General] [answered]
I'm trying to install moose in a multiuser environment on a HPC.  I've plenty of experience with installing applications on the HPC but I'm new to moose. I am able to get it working no problem but am unable to see how to set it up how I wanted.  So its less a technical question about specific details of the install and more a broader question about whether the approach I'm trying makes sense given the way moose is used, and if not, what do other people do in this situation.

I've worked though the "Multi-User Cluster Instructions" here https://mooseframework.inl.gov/getting_started/installation/cluster.html but they kind of peter out before they get to the point I'm having problems with

So broadly the idea is that we have installations of moose that are snapshots in time, so they are not updated with git, if we want an up to date install we create a new dated installation to sit alongside the existing.

And each installation of moose should be useable by different people for different projects.

We're not using conda - a quick over view of the install is
install gcc9.2 / mpich / cmake (shared between all the installations of moose)
git clone https://github.com/idaholab/moose.git into a shared location
./scripts/update_and_rebuild_petsc.sh
./scripts/update_and_rebuild_libmesh.sh
cd test && make -j8

Now originally when we decided on this approach it was thought that building a project against moose didn't alter the moose directory itself.  However I've realised now that the moose installation does get altered which calls into question this strategy.

So now I'm thinking each group / project will need to have its own moose directory?  Is that the way the way people typically work?  Is it possible to have moose installed in a way that people build different projects agains a single install of moose?  And if not what do people typically do instead in HPC / multi-user environment.

Thanks

**Comment by aeslaughter** [comment, marked as answer]
> Now originally when we decided on this approach it was thought that building a project against moose didn't alter the moose directory itself. However I've realised now that the moose installation does get altered which calls into question this strategy.

What is getting altered? That should not be the case?

**Reply by milljm** [reply]
Building a MOOSE based application requires that that user have write access to the moose repository. This is due to object files being created during that user's application build (`export MOOSE_DIR=/shared/location/moose; cd my_app; make`).

The `make` or 'build' process needs to create files/directories in the MOOSE_DIR location during build. This is why we recommend each individual user maintain their own copy of MOOSE.

There are _some_ things you can share. PETSc, and libMesh. However, I would recommend against it, as that will put great strain on the system admin (if that is you); Adding new iterations of these two libraries and keeping them in sync with every iteration of MOOSE your end users are wanting to develop against. This creates many iterations in the long run. This was the reason we introduced the two scripts: `update_and_rebuild_petsc.sh; update_and_rebuild_libmesh.sh`. They are designed to be executed by the end user (easily).

The bottom line, as a developer that developer will need write access to the framework (MOOSE repository).

If you have end users wanting to instead _execute_ a MOOSE based app, that is very much sharable. As a working example using [Pika](https://github.com/idaholab/pika), this is what I mean:

```bash
su user_with_write_access
umask 022
cd /some/shared/location
git clone https://github.com/idaholab/pika
cd pika
git submodule update --init
cd moose/scripts
./scripts/update_and_rebuild_petsc.sh
./scripts/update_and_rebuild_libmesh.sh
cd ../../
make
```

[Pika](https://github.com/idaholab/pika) (a MOOSE based application) can now be run by anyone, so long as the shared location is readable/executable.

```bash
su johndoe
/some/shared/location/pika/pika-opt --help
```

I welcome input from the reset of the team (I was asked to answer this), so it stands to reason they may wish to avoid requiring write access in the moose repository when building other apps.
